compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/descriptors/JvmDeclarationFactory.kt

```diff
@@ -152,6 +155,37 @@ class JvmDeclarationFactory(
         else
             getFieldForObjectInstance(singleton)
 
+    private fun IrClass.allFieldsAreJvmField() =
+        declarations.all { it !is IrProperty || it.backingField?.hasAnnotation(JvmAbi.JVM_FIELD_ANNOTATION_FQ_NAME) == true }
+
+    // We don't move fields to interfaces unless all fields are annotated with @JvmField.
+    // It is an error to annotate only some of the fields of an interface companion with @JvmField.
+    fun getStaticBackingFieldParent(irClass: IrClass): IrClass =
+        if (irClass.isCompanion && (!irClass.parentAsClass.isJvmInterface || irClass.allFieldsAreJvmField()))
+            irClass.parentAsClass
+        else
+            irClass
+
+    fun getStaticBackingField(irProperty: IrProperty, cachedNewParent: IrClass? = null): IrField? =
+        staticBackingFields.getOrPut(irProperty) {
+            if (irProperty.origin == IrDeclarationOrigin.FAKE_OVERRIDE) return null
+            val oldField = irProperty.backingField ?: return null
+            val oldParent = irProperty.parent as? IrClass ?: return null
+            if (!oldParent.isCompanion && !oldParent.isObject) return null
```

Comment:
I'd prefer if these simple conditions were checked outside the `getOrPut` call, to avoid recording nulls in this map for every property in the program

Reply:
Note the `return` - nulls are not actually stored. Yeah, this is indeed confusing and I should not have done that.